README.md
# Overview [![Build Status](https://travis-ci.com/channelaccess/ca.svg?branch=master)](https://travis-ci.com/channelaccess/ca) [![codecov](https://codecov.io/gh/channelaccess/ca/branch/master/graph/badge.svg)](https://codecov.io/gh/channelaccess/ca) ![Maven Central](https://img.shields.io/maven-central/v/org.epics/ca) [ ![JCenter](https://api.bintray.com/packages/paulscherrerinstitute/hltools/org.epics%3Aca/images/download.svg) ](https://bintray.com/paulscherrerinstitute/hltools/org.epics%3Aca/_latestVersion)
The __ca__  library is a pure Java Channel Access client implementation. __ca__ strives to provide the _easiest_ 
way of accessing EPICS Channel Access channels from Java.

The notes in this README file are aimed to assist the __users__ of the library. Developers and __maintainers__ of the library are 
invited to consult the [Developer](DEVELOPER.md) notes.

The version history and latest changes to the library are described in the [CHANGELOG](CHANGELOG.md). 

The results of the functional tests that have been performed on the latest release are available in the 
[INTEGRATION_TESTS](INTEGRATION_TESTS.md) document.

An extended usage [Example](src/main/java/org/epics/ca/examples/Example.java) is available.

We warmly invite and appreciate __contributions__, __improvement suggestions__ and/or __feedback__ on your experiences with this library. Please use the GitHub [issue tracker](https://github.com/channelaccess/ca/issues), or feel free to [contact us](#Contact) directly.

## Features
* Simplicity.
* Use of Java type system.
* Synchronous and asynchronous operations for get, put, connect.
* Efficient handling of parallel operations without the need to use threads.
* Chaining of actions/operations, e.g. set this, then set that, ...
* Easily get additional metadata to value: Timestamp, Alarms, Graphic, Control.
* Supports channel monitoring, including mask capability to select events-of-interest (eg value, log, alarm and/or 
property changes).
* Supports the following listeners: ConnectionListener, AccessRightsListener.
 * Library can be used inside [MATLAB](https://ch.mathworks.com/discovery/what-is-matlab.html) to provide access to EPICS control system data. (See the [README_MATLAB](README_MATLAB.md) file for further details.)

## Compatibility

Implements CA protocol specified in __Version 4.13__ of the Channel Access Protocol Specification. 

The CA protocol specification is available on the new EPICS website [here](https://docs.epics-controls.org/en/latest/specs/ca_protocol.html).
Because links have a tendency to go out-of-date a [PDF](docs/Channel%20Access%20Protocol%20Specification.pdf) 
version is archived in the [documents](docs) area of this project. 

The __ca__ library 1.x.y series of releases are compatible with __Java 8__ and higher.

It is anticipated that the current __ca__ release range (1.3.x) will be the __last release to support Java 8__ and 
that future releases will require at least __Java 11__.

## Distribution

The __ca__ library is available as an Apache Maven package for use in Java projects which use Maven, Gradle, Ivy etc
as their build system.
 
The current distribution home of the library is [JCenter](https://bintray.com/paulscherrerinstitute/hltools/org.epics%3Aca).
Here you can find the most recent library builds. Earlier versions (up to 1.2.2) were available at 
[Maven Central](https://search.maven.org/artifact/org.epics/ca) and for a while longer (2020-06-27) this practice will be continued.

#### Maven Project Usage

Add the following dependency:

```xml
<dependency>
  <groupId>org.epics</groupId>
  <artifactId>ca</artifactId>
  <version>1.3.2</version>
</dependency>
```

and the following repository:

```xml
<repositories>
    <repository>
      <id>jcenter</id>
      <url>https://jcenter.bintray.com/</url>
    </repository>
</repositories>
```

#### Gradle Project Usage

Add the following dependency:

```gradle
compile 'org.epics:ca:1.3.2'
```

and the following repository:

```gradle
repositories {
    maven {
        url "https://jcenter.bintray.com" 
    }
```

# Library Usage

## Configuration

#### Library Configuration

The __ca__ library supports the following configuration variables which can be specified, once per JVM instance, by 
means of environmental variables set in the host OS, or through Java System properties passed on the command-line.

| Property Name               | Description                                                                             | Default Value |
|-----------------------------|-----------------------------------------------------------------------------------------|---------------|
| CA_LIBRARY_LOG_LEVEL        | The level at which CA library log messages will be sent to the standard output stream.  | "INFO"        | 
| CA_REPEATER_LOG_LEVEL       | The level at which CA Repeater log messages will be sent to the standard output stream. | "INFO"        | 
| CA_REPEATER_DISABLE         | Whether the CA library will start/stop a local CA Repeater instance.                    | "false"       |
| CA_REPEATER_OUTPUT_CAPTURE  | Whether to capture the output of the CA Repeater's log messages.                        | "false"       |
| CA_MONITOR_NOTIFIER_IMPL    | The configuration of the CA library monitor notification engine.                        | see below     |


#### EPICS Channel-Access Protocol Configuration

The __ca__ library supports the configuration of the operating parameters of the EPICS Channel Access Protocol by
means of the "normal" configuration variables. These can be set once per context instance.

These can be specified by means of environmental variables in the host OS, through Java System properties, or by 
passing an appropriately configured properties object when creating a context. The following properties are
supported:

| Property Name            | Description                                                                                | Default Value |
|--------------------------|--------------------------------------------------------------------------------------------|---------------|
| EPICS_CA_ADDR_LIST       | The address list to be used when searching for channels.                                   | empty         | 
| EPICS_CA_AUTO_ADDR_LIST  | Automatically build up search address list by introspecting local network interfaces.      | "true"        | 
| EPICS_CA_CONN_TMO        | The UDP Beacon Message Timeout.                                                            | "30s"         |
| EPICS_CA_REPEATER_PORT   | The port to be used when communicating with the local CA Repeater instance.                | "5065"        |
| EPICS_CA_SERVER_PORT     | The port to be used when broadcasting channel search requests to the CA servers.           | "5064"        | 
| EPICS_CA_MAX_ARRAY_BYTES | The maximum size in bytes of an array/waveform.                                            | unlimited     |

_Note:_ In contrast to other Channel Access libraries EPICS_CA_MAX_ARRAY_BYTES is set to unlimited by default. Usually 
there is no reason to set this property since the memory is dynamically acquired as required.

By default the __ca__ library will discover the available EPICS channels by broadcasting search requests on all 
locally-enabled network interfaces.


#### Monitor Notification Engine Configuration

Internally the __ca__ library uses a monitor notification engine to deliver the notifications received from the remote IOCs 
to the local library users via the Java ```Consumer``` interface.

The ```CA_MONITOR_NOTIFIER_IMPL``` property can be used to configure the properties of this engine as follows:

| Configuration String                                                           | Default Buffer Size     | Default Number of Consumer Notification Threads| Additional Comment |
|--------------------------------------------------------------------------------|-------------------------|------------------------------------------------|----------------------------------------------------|
| "BlockingQueueMultipleWorkerMonitorNotificationServiceImpl {,threads}{,bufsiz}"| Integer.MAX_VALUE       | 16                                             | Threads and buffer size are configurable.          |
| "BlockingQueueSingleWorkerMonitorNotificationServiceImpl {,threads}{,bufsiz}"  | Integer.MAX_VALUE       |  1                                             | Threads parameter is ignored and fixed to 1.       |                                           | Experimental. Attempts to improve on the old one.  |
| "StripedExecutorServiceMonitorNotificationServiceImpl {,threads}"              | Integer.MAX_VALUE       | 10                                             | Uses Heinz Kabbutz StripedExecutorService.         |

Note: 

1. The configuration of the monitor notification engine was always intended as an experimental feature. The LMAX-Disruptor
based notification engines which were available previously have now been retired. The primary reason for this is because 
they did not scale well to multiple channels. 
1. The ```BlockingQueueMultipleWorkerMonitorNotificationServiceImpl``` and ```StripedExecutorServiceMonitorNotificationServiceImpl```
notification engines provide optional configuration parameters allowing the size of the notification buffer and number 
of consumer notification threads to be configured. 
1. In the future it is likely that the ```StripedExecutorServiceMonitorNotificationServiceImpl``` will also be retired 
and that the __ca__ library will offer only a single notification engine based on the blocking queue implementation.  
This engine will remain fully configurable to meet the needs of all client applications.   
1. Further details on the requirements for the monitor notification engine and its performance are available in the
   following [MONITOR_INFO.md](https://github.com/channelaccess/ca/blob/master/MONITOR_INFO.md) file.

## Context

In order to create channels a Context must first be created. The context acts as a channel container. When the context 
is closed, all the channels created with that context will also be automatically closed.

This is how to create a context:

```
Context context = new Context()
```

Each context supports the EPICS Channel-Access configuration properties defined in the previous section. The default properties may be overridden by variables set in the host operating system environment, or by setting Java system properties in the CA library environment. If finer-grained control is required over individual contexts then an appropriately configured properties object can be supplied in the class constructor.

```
Properties properties = new Properties();
properties.setProperty( Context.ProtocolConfiguration.EPICS_CA_ADDR_LIST.toString(), "10.10.10.255");
new Context(properties);
```

The context's resources (= network sockets and allocated memory) are disposed of when the context is no longer 
required.  This can be achieved by calling the `close` method.

```
context.close();
```

Alternatively, as the Context implements the Java `AutoCloseable` interface, it can also be used inside a 
try-with-resources statement to ensure that the context's resources are automatically disposed when the 
context goes out of scope. 

```
try ( Context context = new Context)
{
  // Code using context
}
```

## Channel
To create a channel use:

```
Channel<Double> channel = context.createChannel("MY_CHANNEL", Double.class);
```

At channel creation time the channel's type needs to be defined. If you want to have a generic type of channel (i.e. you want to use the type set on the server) use:

```
Channel<Object> channel = context.createChannel("ARIDI-PCT:CURRENT", Object.class);
```

When getting a value from the channel you will get the correct/corresponding Java type that maps to the type set on the server.

After creating the channel object the channel needs to be connected. There is a synchronous and asynchronous way to do so. The synchronous/blocking way is to call:

```
channel.connect()
```

The asynchronous way is to call:

```
`connectAsync()`
```

`connectAsync()` will return a CompletableFuture. To check whether the connect was successful call `.get()` on it. The synchronous way to connect will block until the channel can be connected. If you want to specify a timeout for a connect use the asynchronous connect as follows:

```
channel.connectAsync().get(1, java.util.concurrent.TimeUnit.SECONDS);
```

For connecting multiple channels in parallel use:

```
Channel<Integer> channel1 = context.createChannel("adc02", Integer.class);
Channel<String> channel2 = context.createChannel("adc03", String.class);

// Wait for all channels to be connected
CompletableFuture.allOf(channel1.connectAsync(), channel2.connectAsync()).get();
```

A timeout for the multiple connect is realized the same way as with the single `connectAsync()`.


### Get / Put
After creating a channel you are able to get and put values via the `get()` and `put(value)` methods.

To put a value in a fire and forget style use `putNoWait(value)`. This method will put the value change request on the network but does not wait for any kind of acknowledgement.

```
// Get value
double value = channel.get();
// Set value
channel.put(10.0);
// Set value (best effort style)
channel.putNoWait(10.0);
```

Beside the synchronous (i.e. blocking until the operation is done) versions of `get()` and `put(value)` there are also asynchronous calls. They are named `getAsync()` and `putAsync(value)`. Both functions immediately return with a CompletableFuture for the operation. The Future can be used to wait at any location in the application and to wait for the completion of the operation and to retrieve the final value of the channel.

Example asynchronous get:

```
CompletableFuture<Double> future = channel.getAsync();
CompletableFuture<Double> future2 = channel2.getAsync();

// do something different ...
doSomething();
// ... or simply sleep ...
Thread.sleep(1000);
// ... or simply do nothing ...


double value = future.get();
double value2 = future2.get();
```

Example asynchronous put:

```
CompletableFuture<Status> future = channel.putAsync(1.0); // this could, for example start some move of a motor ...
CompletableFuture<Status> future2 = channel2.putAsync(5.0);

/ do something different ...
doSomething();
// ... or simply sleep ...
Thread.sleep(1000);
// ... or simply do nothing ...


future.get(); // this will return a status object that can be queried if put was successful
future2.get(); // this will return a status object that can be queried if put was successful                                                                                                                                                                                                                                                            
```

### Metadata
If you want to retrieve more metadata besides the value from the channel you can request this by specifying the type of metadata with the get call. For example if you also want to get the value modification/update time besides the value from the cannel use:

```
channel.get(Timestamped.class)
```

Ca supports all metadata types Channel Access provides, namely `Timestamped`, `Alarm`, `Graphic` and `Control`.

|Metadata Type | Metadata                   |
|--------------|----------------------------|
| Timestamped  | seconds, nanos             |
| Alarm        | alarmStatus, alarmSeverity |
| Graphic      | alarmStatus, alarmSeverity, units, precision, upperDisplay, lowerDisplay, upperAlarm, lowerAlarm, upperWarning, lowerWarning |
| Control      | alarmStatus, alarmSeverity, units, precision, upperDisplay, lowerDisplay, upperAlarm, lowerAlarm, upperWarning, lowerWarning, upperControl, lowerControl |

### Monitors
If you want to monitor a channel you can attach a monitor to it like this:

```
Monitor<Double> monitor = channel.addValueMonitor(value -> System.out.println(value));
```

To close a monitor use:

```
monitor.close()
```

Again if you would like to obtain metadata from the monitor you can specify the type of metadata that you are 
interested in.

```
Monitor<Timestamped<Double>> monitor =
    channel.addMonitor(
        Timestamped.class,
            value -> { if (value != null) System.out.println(new Date(value.getMillis()) + " / " + value.getValue()); }
            );
```

### Listeners
A channel can have Access Right and Connection listeners. These two types of listeners are attached as follows.


```
Listener connectionListener = channel.addConnectionListener((channel, state) -> System.out.println(channel.getName() + " is connected? " + state));

Listener accessRightListener = channel.addAccessRightListener((channel, rights) -> System.out.println(channel.getName() + " is rights? " + rights));
```
To remove the listener(s), one can use `try-catch-resources` (i.e. Listeners implement `AutoCloseable`) or

```
listener.close()
```

_Note:_ These listeners can be attached to the channel before connecting.


### ConnectionState
The channels connection state can be checked as follows:

```
channel.getConnectionState()
```

## Channels
The utility class `Channels` provides various convenience functions to perform bulk operations on groups of channels. 

### Create
To create channels `Channels` provides these functions:

```
// Create and connect channel
Channel<String> channel1 = Channels.create(context, "name", String.class);

// Create and connect channel
Channel<String> channel2 = Channels.create(context, new ChannelDescriptor<String>("name", String.class));

// Create and connect multiple channels at once
List<ChannelDescriptor<?>> descriptors = new ArrayList<>();
descriptors.add(new ChannelDescriptor<String>("name", String.class));
descriptors.add(new ChannelDescriptor<Double>("name_double", Double.class));
List<Channel<?>> channels = Channels.create(context,  descriptors);
```

All of these function will __create__ and __connect__ the specified channels.

### WaitForValue
For waiting until a channel reaches a specified value `Channels` provide following functions:

```
waitForValue(channel, "value")

// Use custom comparator for checking what is equal ...
Comparator<String> comparator = ...
waitForValue(channel, "value", comparator)
```

Both functions are also available in an __async__ version. Instead of blocking they return a CompletableFuture.

```
CompletableFuture<String> future = waitForValue(channel, "value")
// ... do something\
future.get();

// Use custom comparator for checking what is equal ...
Comparator<String> comparator = ...
CompletableFuture<String> future1 = waitForValue(channel, "value", comparator)
// ... do something
future1.get()
```

## Annotations
Ca provides the annotation, __@CaChannel__,  to annotate channel declarations within a class. While using the `Channels` utility class these annotations can be used to easily and efficiently create these channels.

All that needs to done is, to annotate the channel declarations as follows:
```
class AnnotatedClass {
	@CaChannel(name="adc01", type=Double.class)
	private Channel<Double> doubleChannel;

	@CaChannel(name="adc01", type=String.class)
	private Channel<String> stringChannel;

	@CaChannel(name={"adc01", "simple"}, type=String.class)
	private List<Channel<String>> stringChannels;

	public Channel<Double> getDoubleChannel() {
		return doubleChannel;
	}
	public Channel<String> getStringChannel() {
		return stringChannel;
	}
	public List<Channel<String>> getStringChannels() {
		return stringChannels;
	}
}	
```

Afterwards the channels can be created via `Channels` as follows:

```
AnnotatedClass object = new AnnotatedClass();
Channels.create(context, object);
```

To close all annotated channels use:

```
Channels.close(object);
```

As channel names should not be hardcoded within an annotation, the name of a channel may contain multiple macros (e.g. `@CaChannel(name="adc${MACRO1}", type=String.class)`). While creating the channels a map of macros need to be passed to the `Channels.create` function.

```
Map<String,String> macros = new HashMap<>();
macros.put("MACRO1","01");
AnnotatedClass object = new AnnotatedClass();
Channels.create(context, object, macros);
```

Macro names are __case sensitive__!

## Examples

Create simple channel:

```
try (Context context = new Context())
{
  try(Channel<Double> channel = Channels.create(context, "MY_CHANNEL", Double.class)){
    System.out.println(channel.get());
  }
}
```

An extended usage example can be found at [src/main/java/org/epics/ca/examples/Example.java](src/main/java/org/epics/ca/examples/Example.java).

# Library API Documentation

Since __ca__ release 1.3.x the Javadoc for the library is published in the GitHub repository [pages](https://channelaccess.github.io/ca/) area.
 
_Note:_ the javadoc is currently in a rather rudimentary state. Over time we plan to improve it. 

# Developer / Maintainer Information

Please see the separate [DEVELOPER](DEVELOPER.md) notes.

# Contact

If you have questions please contact: 'simon.rees@psi.ch' or 'simon.ebner@psi.ch'.


CHANGELOG.md
# Overview

This log describes the functionality of tagged versions within the repository.

# Tags  
* [1.1.0-RELEASE](https://github.com/channelaccess/ca/releases/tag/1.1.0) Released 2018-07-12
  ###### Overview:
  - This release provides an initial fix for the scalability bottleneck which prevented the use of more than ~4000 monitors
    in the Disruptor implementation. The default monitor implementation now uses a single shared work queue from which ten 
    threads take off tasks. The chosen implementation is configurable via a property ('CA_MONITOR_NOTIFIER') which may be 
    set differently for each context or globally. The following values are supported:
      - MultipleWorkerBlockingQueueMonitorNotificationServiceImpl  <<< DEFAULT
      - SingleWorkerBlockingQueueMonitorNotificationServiceImpl
      - DisruptorMonitorNotificationServiceOldImpl"
      - DisruptorMonitorNotificationServiceNewImpl
  - Additionally, various cleanups have been made with the intention of making this library more maintainable in the future.
  ###### Change List: 
  - [Issue #23](https://github.com/channelaccess/ca/issues/23) Documentation Improvements.
     - added comment to README file on CA version compatibility. 
     - added PDF of CA Protocol Specification. Source is [here](https://epics.anl.gov/base/R3-16/1-docs/CAproto/index.html). 
     - added Enterprise Architect Model which can be useful for analysing/documenting the behaviour of this library. 
     - added additional logging (but no functional differences).         
  - [Issue #24:](https://github.com/channelaccess/ca/issues/24)  Upgraded for compatibility with latest versions of Java, Gradle, Disruptor, JUnit etc
     - adjusted build system for Junit 5. 
     - added Mockito framework to support unit testing of TCPTransport.
  - [Issue #25](https://github.com/channelaccess/ca/issues/25)  Fixed bug in TCPTransport receive buffer processing. 
     - added Unit Test which fails on the implementation before the fix. 
  - [Issue #26](https://github.com/channelaccess/ca/issues/26)  Removed TAB character from source files.
  - [Issue #27](https://github.com/channelaccess/ca/issues/27)  Eliminated Javadoc warnings. 
  - [Issue #28](https://github.com/channelaccess/ca/issues/28)  Fix monitor scalability bottleneck.
     - provided support for configurable MonitorNotificationService. 
     - added some additional performance tests.
     - added new test to explore socket handling performance. 
     - removed Disruptor from monitor interface.
     - removed queue zize from monitor interface. 
  - [Issue #29](https://github.com/channelaccess/ca/issues/29) Create software release 1.1.0-RELEASE. 
     - fixed bug in unit test which could result in test raising NPE. 
     - added parameterized unit testing for monitors.    
  - [Issue #30](https://github.com/channelaccess/ca/issues/30) Upgrade to Junit5.
  - [Issue #31](https://github.com/channelaccess/ca/issues/31) Adapt project structure so that 'src/test/java' tree is congruent with 'src/main/java' tree.
  - [Issue #32](https://github.com/channelaccess/ca/issues/32) Add support for uploading builds to PSI artifactory.
  - [Issue #33](https://github.com/channelaccess/ca/issues/33) Cleanup warnings detected by IntelliJ.
  - [Issue #34](https://github.com/channelaccess/ca/issues/34) Provide direct access to socket ByteBuffer through MonitorNotificationService interface.
  - [Issue #35](https://github.com/channelaccess/ca/issues/35) Revert Java buildCompatibility target to Java 8.
  - [Issue #36](https://github.com/channelaccess/ca/issues/35) Revert to java.util.logging throughout.
  
  
* [1.2.0](https://github.com/channelaccess/ca/releases/tag/1.2.0) Released 2018-07-30
   ###### Overview:
   * Provides configurability of the monitor -> consumer notification engine so that the user can choose between a number of implementations including,
      * BlockingQueueMultipleWorkerMonitorNotificationServiceImpl (default)
      * BlockingQueueSingleWorkerMonitorNotificationServiceImpl
      * DisruptorOldMonitorNotificationServiceImpl
      * DisruptorNewMonitorNotificationServiceImpl
      * StripedExecutorServiceMonitorNotificationServiceImpl
   * Changes the default monitor notification engine so that it buffers notification messages to slow consumers rather than throwing them away.
   * Now reports notification buffer overruns to stdout rather than silently throwing them away.
   * Fixes an issue which could occasionally occur with notification messages being delivered out of sequence.
   * Improves the monitor documentation available in the [README](https://github.com/channelaccess/ca/blob/master/README.md) file and in the [MONITOR_INFO](https://github.com/channelaccess/ca/blob/master/MONITOR_INFO.md) file.
   * Provides documentation on measured library performance. See the [MONITOR_INFO](https://github.com/channelaccess/ca/blob/master/MONITOR_INFO.md) file.
   * Increases the test coverage (includes more unit tests; switched to parameterized testing).
   ###### Change List:    
   - [Issue #37](https://github.com/channelaccess/ca/issues/37) Investigate and where necessary fix bug suggested by Fabian with respect to out of order notifications.
   - [Issue #39](https://github.com/channelaccess/ca/issues/39) Provide the ability for library users to configure the appropriate monitor notification strategies. 
   - [Issue #40](https://github.com/channelaccess/ca/issues/40) Decide on whether to keep possibility to send special monitor values (currently null) when a channel disconnects. 
   - [Issue #41](https://github.com/channelaccess/ca/issues/41) Write a monitor notification service impl that buffers to ensure that notifications are not dropped (eg suitable for archiver)
   - [Issue #42](https://github.com/channelaccess/ca/issues/42) Create software release 1.2.0. 
   - [Issue #43](https://github.com/channelaccess/ca/issues/43) Add support for StripedExecutorService. 
   

src/main/java/org/epics/ca/Status.java
package org.epics.ca;


public enum Status
{
   NORMAL (0, Severity.SUCCESS, "Normal successful completion"),
   MAXIOC (1, Severity.ERROR, "Maximum simultaneous IOC connections exceeded"),
   UKNHOST (2, Severity.ERROR, "Unknown internet host"),
   UKNSERV (3, Severity.ERROR, "Unknown internet service"),
   SOCK (4, Severity.ERROR, "Unable to allocate a new socket"),
   CONN (5, Severity.WARNING, "Unable to connect to internet host or service"),
   ALLOCMEM (6, Severity.WARNING, "Unable to allocate additional dynamic memory"),
   UKNCHAN (7, Severity.WARNING, "Unknown IO channel"),
   UKNFIELD (8, Severity.WARNING, "Record field specified inappropriate for channel specified"),
   TOLARGE (9, Severity.WARNING, "The requested transfer is greater than available memory or EPICS_CA_MAX_ARRAY_BYTES"),
   TIMEOUT (10, Severity.WARNING, "User specified timeout on IO operation expired"),
   NOSUPPORT (11, Severity.WARNING, "Sorry, that feature is planned but not supported at this time"),
   STRTOBIG (12, Severity.WARNING, "The supplied string is unusually large"),
   DISCONNCHID (13, Severity.ERROR, "The request was ignored because the specified channel is disconnected"),
   BADTYPE (14, Severity.ERROR, "The data type specifed is invalid"),
   CHIDNOTFND (15, Severity.INFO, "Remote Channel not found"),
   CHIDRETRY (16, Severity.INFO, "Unable to locate all user specified channels"),
   INTERNAL (17, Severity.FATAL, "Channel Access Internal Failure"),
   DBLCLFAIL (18, Severity.WARNING, "The requested local DB operation failed"),
   GETFAIL (19, Severity.WARNING, "Could not perform a database value get for that channel"),
   PUTFAIL (20, Severity.WARNING, "Could not perform a database value put for that channel"),
   ADDFAIL (21, Severity.WARNING, "Could not perform a database monitor add for that channel"),
   BADCOUNT (22, Severity.WARNING, "Count requested inappropriate for that channel"),
   BADSTR (23, Severity.ERROR, "The supplied string has improper format"),
   DISCONN (24, Severity.WARNING, "Virtual circuit disconnect"),
   DBLCHNL (25, Severity.WARNING, "Identical process variable name on multiple servers"),
   EVDISALLOW (26, Severity.ERROR, "The CA routine called is inappropriate for use within an event handler"),
   BUILDGET (27, Severity.WARNING, "Database value get for that channel failed during channel search"),
   NEEDSFP (28, Severity.WARNING, "Unable to initialize without the vxWorks VX_FP_TASK task option set"),
   OVEVFAIL (29, Severity.WARNING, "Event queue overflow has prevented first pass event after event add"),
   BADMONID (30, Severity.ERROR, "bad monitor subscription identifier"),
   NEWADDR (31, Severity.WARNING, "Remote channel has new network address"),
   NEWCONN (32, Severity.INFO, "New or resumed network connection"),
   NOCACTX (33, Severity.WARNING, "Specified task isnt a member of a CA context"),
   DEFUNCT (34, Severity.FATAL, "Attempt to use defunct CA feature failed"),
   EMPTYSTR (35, Severity.WARNING, "The supplied string is empty"),
   NOREPEATER (36, Severity.WARNING, "Unable to spawn the CA repeater thread- auto reconnect will fail"),
   NOCHANMSG (37, Severity.WARNING, "No channel id match for search reply- search reply ignored"),
   DLCKREST (38, Severity.WARNING, "Reseting dead connection- will try to reconnect"),
   SERVBEHIND (39, Severity.WARNING, "Server (IOC) has fallen behind or is not responding- still waiting"),
   NOCAST (40, Severity.WARNING, "No internet interface with broadcast available"),
   BADMASK (41, Severity.ERROR, "The monitor selection mask supplied is empty or inappropriate"),
   IODONE (42, Severity.INFO, "IO operations have completed"),
   IOINPROGESS (43, Severity.INFO, "IO operations are in progress"),
   BADSYNCGRP (44, Severity.ERROR, "Invalid synchronous group identifier"),
   PUTCBINPROG (45, Severity.ERROR, "Put callback timed out"),
   NORDACCESS (46, Severity.WARNING, "Read access denied"),
   NOWTACCESS (47, Severity.WARNING, "Write access denied"),
   ANACHRONISM (48, Severity.ERROR, "Sorry, that anachronistic feature of CA is no longer supported"),
   NOSEARCHADDR (49, Severity.WARNING, "The search/beacon request address list was empty after initialization"),
   NOCONVERT (50, Severity.WARNING, "Data conversion between client's type and the server's type failed"),
   BADCHID (51, Severity.ERROR, "Invalid channel identifier"),
   BADFUNCPTR (52, Severity.ERROR, "Invalid function pointer"),
   ISATTACHED (53, Severity.WARNING, "Thread is already attached to a client context"),
   UNAVAILINSERV (54, Severity.WARNING, "No support in service"),
   CHANDESTROY (55, Severity.WARNING, "User destroyed channel"),
   BADPRIORITY (56, Severity.ERROR, "Priority out of range"),
   NOTTHREADED (57, Severity.ERROR, "Preemptive callback not enabled - additional threads may not join"),
   ARRAY16KCLIENT (58, Severity.WARNING, "Client's protocol revision does not support transfers exceeding 16k bytes"),
   CONNSEQTMO (59, Severity.WARNING, "Virtual circuit connection sequence aborted"),
   UNRESPTMO (60, Severity.WARNING, "Virtual circuit connection unresponsive");

   private final int value;
   private final Severity severity;
   private final String message;

   Status( int value, Severity severity, String message )
   {
      this.value = value;
      this.severity = severity;
      this.message = message;
   }

   public int getValue()
   {
      return value;
   }

   public Severity getSeverity()
   {
      return severity;
   }

   public String getMessage()
   {
      return message;
   }

   public boolean isSuccessful()
   {
      return severity == Severity.SUCCESS;
   }

   /* CA Status Code Definitions */

   public final static int CA_M_MSG_NO = 0x0000FFF8;
   public final static int CA_M_SEVERITY = 0x00000007;
   public final static int CA_M_LEVEL = 0x00000003;
   public final static int CA_M_SUCCESS = 0x00000001;
   public final static int CA_M_ERROR = 0x00000002;
   public final static int CA_M_SEVERE = 0x00000004;

   public final static int CA_S_MSG_NO = 0x0D;
   public final static int CA_S_SEVERITY = 0x03;

   public final static int CA_V_MSG_NO = 0x03;
   public final static int CA_V_SEVERITY = 0x00;
   public final static int CA_V_SUCCESS = 0x00;

   /**
    * Get EPICS status code.
    *
    * @return EPICS status code.
    */
   public final int getStatusCode()
   {
      return ((getValue () << CA_V_MSG_NO) & CA_M_MSG_NO)
            | ((getSeverity ().getValue () << CA_V_SEVERITY) & CA_M_SEVERITY);
   }

   /**
    * Get status instance from EPICS status code.
    *
    * @param code EPICS status code.
    * @return status instance from EPICS status code.
    */
   public static Status forStatusCode( int code )
   {
      return forValue ((code & CA_M_MSG_NO) >> CA_V_MSG_NO);
   }

   public static Status forValue( int value )
   {
      Status[] statusLUT = values ();

      if ( value < 0 || value >= statusLUT.length )
         throw new IndexOutOfBoundsException ("invalid status code " + value);

      return statusLUT[ value ];
   }
}
